File: medium/3741-minimum-distance-between-three-equal-elements-ii/solution.java
import java.util.*;

/**
 * 3741. Minimum Distance Between Three Equal Elements II
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public int minimumDistance(int[] nums) {
        Map<Integer, List<Integer>> indices = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            indices.computeIfAbsent(nums[i], k -> new ArrayList<>()).add(i);
        }

        int minDist = Integer.MAX_VALUE;

        for (List<Integer> idxList : indices.values()) {
            if (idxList.size() >= 3) {
                for (int i = 0; i < idxList.size() - 2; i++) {
                    int span = idxList.get(i + 2) - idxList.get(i);
                    minDist = Math.min(minDist, 2 * span);
                }
            }
        }

        return minDist == Integer.MAX_VALUE ? -1 : minDist;
    }
}

File: medium/3741-minimum-distance-between-three-equal-elements-ii/solution.py
from typing import List
from collections import defaultdict


class Solution:
    def minimumDistance(self, nums: List[int]) -> int:
        """
        3741. Minimum Distance Between Three Equal Elements II
        Time: O(n)
        Space: O(n)
        """
        indices = defaultdict(list)

        for i, num in enumerate(nums):
            indices[num].append(i)

        min_dist = float('inf')

        for idx_list in indices.values():
            if len(idx_list) >= 3:
                for i in range(len(idx_list) - 2):
                    span = idx_list[i + 2] - idx_list[i]
                    min_dist = min(min_dist, 2 * span)

        return min_dist if min_dist != float('inf') else -1
